src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java

```diff
@@ -0,0 +1,382 @@
+                }
+
+                @Override
+                public void finishBundle(Context c) throws Exception {
+                    super.finishBundle(c);
+                    final Instant instant = new Date().toInstant();
+                    long ms = Duration.between(started.toInstant(), instant).toMillis();
+                    String q = querying.elapsed(TimeUnit.MILLISECONDS) + " ms";
+                    String l = loading.elapsed(TimeUnit.MILLISECONDS) + " ms";
+                    String ch = choosing.elapsed(TimeUnit.MILLISECONDS) + " ms";
+                    String s = shipping.elapsed(TimeUnit.MILLISECONDS) + " ms";
+                    bunny.stepEnd("Bundle sent " + readsEmitted + " reads in " + ms + " ms. Breakdown: querying " + q + " loading " + l + " choosing " + ch + " shipping " + s);
+                }
+
+                @Override
+                public void processElement(ProcessContext c) throws Exception {
+                    int rejected = 0;
+                    int accepted = 0;
+                    Stopwatch processing = Stopwatch.createStarted();
+
+                    Stopwatch kving = Stopwatch.createUnstarted();
+                    Stopwatch outputting = Stopwatch.createUnstarted();
+
+                    Stopwatch readingInput = Stopwatch.createStarted();
+                    SimpleInterval i = c.element();
+                    readingInput.stop();
+                    Stopwatch init = Stopwatch.createStarted();
+                    ArrayList<SimpleInterval> is = new ArrayList<>();
+                    is.add(i);
+
+                    Stopwatch cutShards = Stopwatch.createStarted();
+                    List<SimpleInterval> shards = cutToShards(is, outputShardSize);
+                    cutShards.stop();
+
+                    int currentShardIndex = 0;
+                    SimpleInterval currentShard = shards.get(currentShardIndex);
+                    List<GATKRead> ret = new ArrayList<GATKRead>();
+                    init.stop();
+
+                    if (!cloudStorageUrl && !hadoopUrl) {
+                        // local file, use setIntervalsForTraversal
+                        querying.start();
+                        ReadsDataSource rds = new ReadsDataSource(new File(bam));
+                        rds.setIntervalsForTraversal(is);
+                        querying.stop();
+                        choosing.start();
+                        for (GATKRead r : rds) {
+                            if (!accept(r, i)) {
+                                rejected++;
+                                continue;
+                            }
+                            while (currentShard.getEnd() < r.getStart()) {
+                                // move to the next shard.
+                                if (!ret.isEmpty()) {
+                                    accepted += ret.size();
+                                    choosing.stop();
+                                    shipping.start();
```

Comment:
Remove stopwatch code from the version going into master, as it makes the class illegible